### PR TITLE
Adjust for Paddle Billing v2

### PR DESCRIPTION
**Pull request by bobbyiliev**
Preparing the upgrade for the Paddle v2 billing changes:

> https://developer.paddle.com/changelog/overview

### CODE REVIEW

**Comment by rafa-acioly** [comment]
Is there a way to use "v1" of paddle billing for now? It seems like paddle have fully migrated to v2

**Comment by bobbyiliev** [comment]
> Is there a way to use "v1" of paddle billing for now? It seems like paddle have fully migrated to v2

I think that what you could do is reach out to them and ask for access to Paddle Classic so you could use the old API version.

**Comment by alfreddagenais** [comment]
Hello @bobbyiliev,

First off, I want to express my gratitude for this Pull Request. Although I understand it's still in draft mode, I found your contribution immensely helpful in adjusting my own code, especially with the significant changes in Paddle's version 2.

I have made several edits to your initial proposal to align more closely with Paddle's documentation, but I want to emphasize that the majority of your work was already very well executed.

Additionally, I'd like to offer my assistance or any ideas if you need them. Please feel free to let me know if I can contribute in any way. Your initial approach has been quite inspiring, and I'd be delighted to collaborate further on this project.

Thank you again for your valuable work on this project.

Best regards,

**Comment by bobbyiliev** [comment]
Hey @alfreddagenais! Thanks for the great feedback! I've postponed this for a bit but planning to pick it backup after the holidays.

Still debating on whether to completely switch to using the Laravel Cashier package. There has been some ongoing development for Paddle v2 but it is still under alpha release:

> https://github.com/laravel/cashier-paddle/releases/tag/v2.0.0-alpha.3

Totally open to suggestions here!

**Comment by alfreddagenais** [comment]
Hey @bobbyiliev !

Thank you for your update and for considering my feedback. I completely understand the need to postpone due to the holiday season.

I wanted to share some insights from my recent experience. Yesterday, while integrating the details from the PR, I noticed there are quite a few aspects that need to be addressed in the code to achieve a more refined version for payments, especially for Wave's consumers/developers.

After thoroughly reviewing the complete documentation of Laravel Cashier (https://laravel.com/docs/10.x/cashier-paddle), I must say I'm really impressed with this package. Notably, it's already advanced to version 2.0.4 (2 days ago... not just the [2.0.0-alpha.3](https://github.com/laravel/cashier-paddle/releases/tag/v2.0.0-alpha.3)). In my opinion, incorporating the Cashier-Paddle package into Wave could be highly beneficial. Given the extensive features it already manages within the Laravel community, it seems counterproductive to rewrite these parts. Moreover, its webhook event management and validation features (https://github.com/laravel/cashier-paddle/releases/tag/v2.0.4) are particularly appealing.

Based on these factors, I have decided to adjust my code to utilize Cashier-Paddle. It simplifies the process significantly, reducing the amount of code and potential sources of error. This is especially crucial considering the evolving nature of payment engines.

I believe leveraging Cashier-Paddle could streamline our development process and enhance the overall functionality of Wave. Looking forward to hearing your thoughts on this!

Best regards,

PS: Just a quick note, for me is missing a issues tab in github wave project. You can track the todo, improvement, etc. And the community looking the the  wave package.

**Comment by podopodo** [comment]
Hi @bobbyiliev 
Thanks for your works in modifying to the new version of Paddle. From the above conversation, not sure if you'll be deciding to do away with Paddle completely. If you do however plan to stick to it,
1) AutoFill during Paddle.Checkout.open in checkout.blade.php
- "email" key doesn't work.
```
Paddle.Checkout.open({
    items: product,
    email: '@if(!auth()->guest()){{ auth()->user()->email }}@endif',
    successCallback: "checkoutComplete",
});
```
- "email" needs to be contained in the key "customer". Source: [https://developer.paddle.com/build/checkout/prefill-checkout-properties](https://developer.paddle.com/build/checkout/prefill-checkout-properties)
```
Paddle.Checkout.open({
    items: product,
    customer: {email: '@if(!auth()->guest()){{ auth()->user()->email }}@endif'},
    successCallback: "checkoutComplete",
});
```

2) "Get Started" or Payment first and then Signup is not working with error from paddle.js
`Uncaught Error: Checkout input failed validation : Error: customer ID or email is required`
when initiating the checkout. Conditionally removing the "customer" object seems to work if its guest.

```
Paddle.Checkout.open({
    items: product,
    successCallback: "checkoutComplete",
});
```

Haven't checked on the consequent implications of these changes yet. I am new to both Laravel and Wave.

File: wave/resources/views/checkout.blade.php [hunk 1]
```diff
@@ -1,11 +1,24 @@
-<script src="https://cdn.paddle.com/paddle/paddle.js"></script>
+<script src="https://cdn.paddle.com/paddle/v2/paddle.js"></script>
```
Comment (l-alexandrov) [inline wave/resources/views/checkout.blade.php:1]:
```suggestion
<script src="https://cdn.paddle.com/paddle/v2/paddle.js" defer></script>
```

File: wave/resources/views/checkout.blade.php [hunk 2]
```diff
@@ -50,9 +71,74 @@ function waveUpdate(){
...
+        }
+    }
+
+    window.checkoutCancel = function(data){
+        let subscriptionId = data.id;
+        axios.post('/cancel', { _token: csrf, id: subscriptionId })
+            .then(function (response) {
+                if(parseInt(response.data.status) == 1){
+                    window.location = '/settings/subscription';
+                }
```
Comment (l-alexandrov) [inline wave/resources/views/checkout.blade.php:141]: Extact this to a dedicated js file and defer its loading.

**Review by l-alexandrov — COMMENTED** [review]
Here are some page speed improvement suggestions.

File: wave/resources/views/checkout.blade.php (hunk 1 above)
Comment (bobbyiliev) [reply wave/resources/views/checkout.blade.php:1]: This does not seem to be working unfortunately:

```
Uncaught ReferenceError: Paddle is not defined
    at pricing:984:5
```

Comment (l-alexandrov) [reply wave/resources/views/checkout.blade.php:1]: Yes, I guess so. I didn't manage to combine my suggestion in a single entry. You need to defer the script after this one too (https://github.com/thedevdojo/wave/pull/100#discussion_r1544362982).
We need to make sure that we try to instantiate Paddle after the paddle script is loaded.

**Review by tnylea — APPROVED** [review]
All looks great to me. Excited to get this merged in 👍

**Pull request merged**